### PR TITLE
Documentation on custom package repositories

### PR DESCRIPTION
**Pull request by jonasfj**
Authentication with tokens is new in 2.15, which sort of makes how to use custom package repositories a bit more interesting.

This was just my attempt at writing something that helps people understand custom package repositories, without having to piece together the different bits from all over. 

Staged: https://kw-staging-dartlang-2.web.app/tools/pub/custom-package-repositories

### CODE REVIEW

File: src/tools/pub/dependencies.md [hunk 1]
```diff
@@ -114,8 +114,8 @@ This example specifies that your package depends on a hosted package named
 `transmogrify` and will work with any version from 1.4.0 to 2.0.0
 (but not 2.0.0 itself).
 
-If you want to use your own package server, you can use a description that
-specifies its URL:
+If you want to use your [own package server](./custom-package-repository),
```
Comment (parlough) [inline src/tools/pub/dependencies.md:117]:
```suggestion
If you want to use your [own package server](/custom-package-repositories),
```

File: src/tools/pub/custom-package-repositories.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+---
+
+The `pub` tool has support for third-party package repositories. A package
+repository is an HTTP server from which the `pub` client can fetch packages.
+The official package repository is [pub.dev](https://pub.dev), this repository
+is by the `pub` tool by default. A package repository is identified by a
+_hosted-url_ (e.g. `https://pub.example.com/my-account/`). 
+
+Third-party package servers can be useful for hosting private packages. It is
+also common to use [git-dependencies](./dependencies#git-packages) for hosting
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:13]:
```suggestion
also common to use [git-dependencies](/dependencies#git-packages) for hosting
```

File: src/tools/pub/custom-package-repositories.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+private packages, however, the `pub` does not support resolving versions against
+a git-repository it just fetches a specific revision of the git-repository.
+Hence, when many people are collaborating it is sometimes preferable to use a
+private package repository.
+
+
+## Authenticating with a custom package repository
+Custom package servers are often private and require authentication. The `pub`
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:21]:
```suggestion
## Authenticating with a custom package repository

Custom package servers are often private and require authentication. The `pub`
```

File: src/tools/pub/custom-package-repositories.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+tool authenticates against package repositories by attaching a secret token to
+the requests. When the `pub` tool does not have a token for a given
+_hosted-url_ it will attempt to make requests without authentication.
+
+To specify a secret token for the `pub` tool to use with a given _hosted-url_,
+use the `dart pub token add <hosted-url>` command.
+
+
+## Getting dependencies from a custom package repository
+To fetch a package from custom package repository we must specify the
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:31]:
```suggestion
## Getting dependencies from a custom package repository

To fetch a package from custom package repository we must specify the
```

File: src/tools/pub/custom-package-repositories.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+_hosted-url_ for the package in `pubspec.yaml`, using the syntax for
+[hosted packages](./dependencies#hosted-packages).
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:40]:
````suggestion
```yaml
dependencies:
  foo:
    hosted: https://pub.example.com
    version: ^1.4.0
```
````

File: src/tools/pub/custom-package-repositories.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+In the example above `package:foo` will be fetched from
+`https://pub.example.com`. If authentication is required by this package
+repository, it can be specified using
+`dart pub token add https://pub.example.com`, which will prompt for the secret
+token.
+
+
+### Using multiple package repositories
+It is possible to fetch different dependencies from different package
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:50]:
```suggestion
### Using multiple package repositories

It is possible to fetch different dependencies from different package
```

File: src/tools/pub/custom-package-repositories.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  # package retry is fetched from pub.dev (the default package repository)
+  retry: ^3.0.0
+  # package foo is fetched from https://pub.example.com
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:62]:
````suggestion
```yaml
dependencies:
  # package retry is fetched from pub.dev (the default package repository)
  retry: ^3.0.0
  # package foo is fetched from https://pub.example.com
  foo:
    hosted: https://pub.example.com
    version: ^1.4.0
```
````

File: src/tools/pub/custom-package-repositories.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+and either update the `dependencies` section in section in each package, or
+override the default package repository.
+
+{{site.alert.note}}
+To ensure that public packages are usable to everyone, the official package
+repository [pub.dev](https://pub.dev) does not allow publication of packages
+with git-dependencies or hosted-dependencies with custom package repositories.
+
+However, such packages can be published to a custom package repository!
+{{site.alert.end}}
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:78]:
```suggestion
{{site.alert.note}}
  To ensure that public packages are usable to everyone, the official package
  repository [pub.dev]({{site.pub}}) does not allow publication of packages
  with git-dependencies or hosted-dependencies with custom package repositories.

  However, such packages can be published to a custom package repository!
{{site.alert.end}}
```

File: src/tools/pub/custom-package-repositories.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+
+### Overriding the default package repository
+The package repository to be used can be specified on a per-dependency basis,
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:82]:
```suggestion
### Overriding the default package repository

The package repository to be used can be specified on a per-dependency basis,
```

File: src/tools/pub/custom-package-repositories.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+or the default package repository can be overriden with the
+[`PUB_HOSTED_URL`](./environment-variables) environment variable.
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:84]:
```suggestion
[`PUB_HOSTED_URL`](/environment-variables) environment variable.
```

File: src/tools/pub/custom-package-repositories.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+This can be useful if you are mirroring all packages in a private package
+repository, or working in a restricted network environment using a mirror of
+[pub.dev](https://pub.dev).
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:88]:
```suggestion
[pub.dev]({{site.pub}}).
```

File: src/tools/pub/custom-package-repositories.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+
+## Publishing to a custom package repository
+The `pub` tool also supports publishing packages to a private package
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:92]:
```suggestion
## Publishing to a custom package repository

The `pub` tool also supports publishing packages to a private package
```

File: src/tools/pub/custom-package-repositories.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+repository, using the [`publish_to`](./pubspec#publish_to) property in
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:93]:
```suggestion
repository, using the [`publish_to`](/pubspec#publish_to) property in
```

File: src/tools/pub/custom-package-repositories.md [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+`pubspec.yaml`. If working on a private package it is a good idea to specify
+this early in the development, so as to prevent accidental publication to
+[pub.dev](https://pub.dev).
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:96]:
```suggestion
[pub.dev]({{site.pub}}). 
```

File: src/tools/pub/custom-package-repositories.md [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+To publish a package to `https://pub.example.com` you would write a
+`pubspec.yaml` as follows, and run `dart pub publish`.
+
+{% prettify yaml tag=pre+code %}
+name: mypkg
+version: 1.0.0
+# Ensures the package is published to https://pub.example.com
+publish_to: https://pub.example.com
+{% endprettify %}
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:106]:
````suggestion
```
name: mypkg
version: 1.0.0
# Ensures the package is published to https://pub.example.com
publish_to: https://pub.example.com
```
````

File: src/tools/pub/custom-package-repositories.md [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+{{site.alert.note}}
+Even if you don't have a private repository, you can specify `publish_to: none`
+which stops accidental publication.
+{{site.alert.end}}
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:111]:
```suggestion
{{site.alert.note}}
  Even if you don't have a private repository, 
  you can specify `publish_to: none`
  which stops accidental publication.
{{site.alert.end}}
```

File: src/tools/pub/custom-package-repositories.md [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+
+## Writing a custom package repository
+The REST API for writing a custom package repository is outlined in the
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:115]:
```suggestion
## Writing a custom package repository

The REST API for writing a custom package repository is outlined in the
```

File: src/tools/pub/custom-package-repositories.md [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+_hosted-url_ it will attempt to make requests without authentication.
+
+To specify a secret token for the `pub` tool to use with a given _hosted-url_,
+use the `dart pub token add <hosted-url>` command.
+
+
+## Getting dependencies from a custom package repository
+To fetch a package from custom package repository we must specify the
+_hosted-url_ for the package in `pubspec.yaml`, using the syntax for
+[hosted packages](./dependencies#hosted-packages).
```
Comment (parlough) [inline src/tools/pub/custom-package-repositories.md:33]:
```suggestion
[hosted packages](/dependencies#hosted-packages).
```

**Review by parlough — COMMENTED** [review]
Some formatting and link suggestions.

File: src/tools/pub/custom-package-repositories.md [hunk 19]
```diff
@@ -0,0 +1,119 @@
+---
+title: Custom Package Repositories
```
Comment (kwalrath) [inline src/tools/pub/custom-package-repositories.md:2]:
```suggestion
title: Custom package repositories
```

**Comment by kwalrath** [comment]
Thanks for the page, @jonasfj! We'll do a copy edit pass once you've addressed other reviewers' feedback.

File: src/tools/pub/custom-package-repositories.md (hunk 2 above)
Comment (jonasfj) [reply src/tools/pub/custom-package-repositories.md:13]: I'm guessing that if we want to make links absolute we have start from `/tools/...`

File: src/tools/pub/custom-package-repositories.md (hunk 5 above)
Comment (jonasfj) [reply src/tools/pub/custom-package-repositories.md:40]: I didn't know we could do that, we should probably patch the other pages -- I just cargo culted.. :D

**Comment by jonasfj** [comment]
Thanks, I think I addressed all comments. Feel free to edit as much as you please :D

I don't know if an entry in the sidebar is warranted, I didn't add it as I wasn't sure it was necessary (or where to put it).

**Review by kwalrath — APPROVED** [review]

**Pull request merged**